dialogue-blocking-channels/src/main/java/com/palantir/dialogue/blocking/DefaultCallingThreadExecutor.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.blocking;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListenableFutureTask;
+import com.palantir.dialogue.RequestAttachmentKey;
+import com.palantir.dialogue.futures.DialogueFutures;
+import com.palantir.logsafe.Preconditions;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.RunnableFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class DefaultCallingThreadExecutor implements CallingThreadExecutor {
+
+    static final RequestAttachmentKey<CallingThreadExecutor> ATTACHMENT_KEY =
+            RequestAttachmentKey.create(CallingThreadExecutor.class);
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultCallingThreadExecutor.class);
+    private static final boolean DO_NOT_INTERRUPT = false;
+    private final long threadId = Thread.currentThread().getId();
+    private final Queue queue;
+
+    @VisibleForTesting
+    interface QueueTake {
+        <E> E take(BlockingQueue<E> queue) throws InterruptedException;
+    }
+
+    @VisibleForTesting
+    DefaultCallingThreadExecutor(QueueTake queueTake) {
+        queue = new Queue(queueTake);
+    }
+
+    DefaultCallingThreadExecutor() {
+        this(BlockingQueue::take);
+    }
+
+    @Override
+    public ListenableFuture<?> submit(Runnable task) {
+        return queue.submit(task);
+    }
+
+    @Override
+    public void executeQueue(ListenableFuture<?> await) {
+        Preconditions.checkState(Thread.currentThread().getId() == threadId, "Executing queue on different thread");
+        await.addListener(() -> queue.submitNotifier(queue::poison), DialogueFutures.safeDirectExecutor());
+        try {
+            RunnableFuture<?> toRun;
+            while ((toRun = queue.getWork()) != null) {
+                toRun.run();
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            abortQueue();
+
+            await.cancel(DO_NOT_INTERRUPT);
+        }
+    }
+
+    private void abortQueue() {
+        queue.poison();
+
+        RunnableFuture<?> toRun;
+        while ((toRun = queue.poll()) != null) {
+            toRun.cancel(DO_NOT_INTERRUPT);
```

Comment:
Note: I'm not sure this does anything because we wrap runnables with FutureTask, and never use the result except to call `cancel()`. Leaving this as is makes the future returned from SameThreadExecutor make a little more sense if we change how we use it, but thar be dragons.

No change requested, just a case that took me a bit to reason through.

Reply:
Ah, I think I understand your previous concern a little bit more now, but still a bit hazy. The futures returned from #submit get transformed further down the line right? so it's very important we;re able to cancel them, so we don't leave resources that are tied to these futures (like concurrency limiters) leaked?

Reply:
I think the futures returned from our executor are exclusively used for cancellation, and never transformed. The adapter submits a runnable which completes a SettableFuture, otherwise we risk leaking responses in cancellation races. There's unfortunately a lot of code in DialogueFutures to prevent resource leaks, which we may be able to replace with CloseableFutures from recent guava versions which didn't exist when this was written.